src/test/functional/functional-tests.js

```diff
@@ -628,98 +626,24 @@ describe('functional tests', function() {
-          done()
-        })
-      })
-    })
 
-    step(`setBucketPolicy(bucketName, objectPrefix, bucketPolicy)_bucketName:${bucketName}, bucketPolicy:READONLY_`, done => {
-      client.setBucketPolicy(bucketName, '', Policy.READONLY)
-        .then(() => done())
-        .catch(done)
-    })
-
-    step(`getBucketPolicy(bucketName, objectPrefix)_bucketName:${bucketName}_`, done => {
-      client.getBucketPolicy(bucketName, '')
-        .then(response => {
-          if (response != Policy.READONLY)
-            return done(new Error(`policy is incorrect (${response} != ${Policy.READONLY})`))
-          done()
-        })
-        .catch(done)
-    })
-
-    step(`setBucketPolicy(bucketName, objectPrefix, bucketPolicy)_bucketName: ${bucketName}, objectPrefix: prefix, bucketPolicy: READWRITE_`, done => {
-      policyNotImplemented = false
-      client.setBucketPolicy(bucketName, 'prefix', Policy.READWRITE, err => {
-        if (err) {
-          if(err.code != 'NotImplemented') return done(err)
-          else policyNotImplemented = true
-        }
-        done()
-      })
-    })
-
-    step(`getBucketPolicy(bucketName, objectPrefix)_bucketName: ${bucketName}, objectPrefix: prefix_`, done => {
-      if (!policyNotImplemented) {
-        client.getBucketPolicy(bucketName, 'prefix')
-          .then(response => {
-            if (response != Policy.READWRITE)
-              return done(new Error(`policy is incorrect (${response} != ${Policy.READWRITE})`))
-            done()
-          })
-          .catch(done)
-      } else done()
-    })
-
-    step(`getBucketPolicy(bucketName, objectPrefix)_bucketName:${bucketName}, objectPrefix: wrongprefix_`, done => {
-      client.getBucketPolicy(bucketName, 'wrongprefix')
-        .then(response => {
-          if (response == Policy.READWRITE)
-            return done(new Error(`policy is incorrect (${response} == ${Policy.READWRITE})`))
-          done()
-        })
-        .catch(done)
-    })
+  describe('bucket policy', () => {
+    let policy = `{"Version":"2012-10-17","Statement":[{"Action":["s3:GetBucketLocation"],"Effect":"Allow","Principal":{"AWS":["*"]},"Resource":["arn:aws:s3:::${bucketName}"],"Sid":""},{"Action":["s3:ListBucket"],"Condition":{"StringEquals":{"s3:prefix":["foo","prefix/"]}},"Effect":"Allow","Principal":{"AWS":["*"]},"Resource":["arn:aws:s3:::${bucketName}"],"Sid":""},{"Action":["s3:GetObject"],"Effect":"Allow","Principal":{"AWS":["*"]},"Resource":["arn:aws:s3:::${bucketName}/foo*","arn:aws:s3:::${bucketName}/prefix/*"],"Sid":""}]}`
 
-    step(`setBucketPolicy(bucketName, objectPrefix, bucketPolicy, cb)_bucketName:${bucketName}, bucketPolicy: NONE_`, done => {
-      client.setBucketPolicy(bucketName, '', Policy.NONE, err => {
+    step(`setBucketPolicy(bucketName, bucketPolicy, cb)_bucketName:${bucketName}, bucketPolicy:${policy}_`, done => {
+      client.setBucketPolicy(bucketName, policy, err => {
+        if (err && err.code == 'NotImplemented') return done()
```

Comment:
Do we need to check for `err` here? 
Is it  possible to have `err.code == 'NotImplemented'` without having an `err`?
If not, then there is no need to check for `err`.